### DISCUSSION TITLE
ATtiny1627 vs ATtiny1626

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [SpenceKonde/megaTinyCore#607] [Q&A]
I had someone create some test code on the ATtiny1627 Nano Board using Microchip Studio. I have a raw ATtiny1626 chip soldered on a breakout board which I've been experimenting on with Arduino and MTC. I have burned the Bootloader onto this chip.

- The code uses TCA0 to generate a 20kHz PWM on PB5. 
- We use PB1 as sort of a scope probe and when it touches PB5 and see the PWM it blinks an LED on PB0.
- If PB1 touches GND then the LED should stop blinking.

This code runs fine on the Nano board but for some reason on my 1626 board when PB1 touches GND to LED signal gets a repeating low pulse of ~50ms at ~1.5Hz. The LED's cathode is connected to PB0 so a high turns it OFF.

All the experimenting I did in Arduino with other code seemed to work as expected and the chip doesn't appear to be damaged.

Maybe the chip is internally resetting for some reason and when it does PB0 goes low for 50ms? Not sure why that would happen but the only thing I can think of is that something I did when I burned the bootloader or assigned reset on PB4 is interfering with the code that Microchip Studio created?

I don't have Microchip Studio loaded on my PC so I can't run it and poke around registers to see what's happening and the chip resetting is only a thought on my part. I just thought I'd ask to see if anyone had any ideas. Short of removing the 1626 from the breakout board and replacing it is there a way to reset everything in the chip to see if somehow this is not a hardware problem?

**Comment by SpenceKonde** [comment]
What is the behavior on the Curiosity Nano, and how does it differ from the 1626? 

Can you please post code that reproduces this issue? I can't comment further without looking over the code.

**Reply by GeorgeIoak** [reply]
The person who wrote the test code didn't send me the source as he's out now. I was hoping to just try a "full reset" of the chip first to even see if that changed the behavior.

The curiosity Nano performs as expected meaning that when PB1 is grounded the LED output (PB0) goes high and the LED turns OFF. If PB1 touches PB5 the LED starts blinking. 

So the LED is OFF in the power up condition and then if it touches PB5 the LED blinks.

**Reply by SpenceKonde** [reply]
that link is 404.

**Reply by GeorgeIoak** [reply]
it works for me but I forked a private repo and I thought that would give you access but apparently not. Let me download it and create a new public repo...

**Comment by SpenceKonde** [comment]
If you don't want it public, you can make a private repo and give me access, too.

**Reply by GeorgeIoak** [reply]
It's not complete so I'll share for now to make it easy, https://github.com/GeorgeIoak/ATtiny1626-Test

Thanks for looking.

**Reply by SpenceKonde** [reply]
still getting 404  x_x

**Reply by GeorgeIoak** [reply]
It made it private again so i added you and you should get an invite, I hope ..

**Reply by SpenceKonde** [reply]
I see variables not declared volatile being written from inside of ISRs and read outside of them, and interrupts are being disabled and reenabled around things that do not need that - or at least that wouldn't if the code handled the ISR-modified-variables correctly, I don't immediately see one place that I can point to as being where the difference in behavior comes from - but the ratio of red flags to code is extremely high. It looks to me like the author of the code is not particularly familiar with the basics of interrupts? at least not on AVR....

**Reply by GeorgeIoak** [reply]
Thanks for taking the time to take a look at the code. I'll have to take a closer look at how my board (with the 1626) functions compared to the 1627 based Nano. Since I was experimenting with Arduino and my board I was concerned that I maybe did something to change the behavior by mistake. Maybe I'll bite the bullet and just put down a fresh 1626 chip on my board to see what happens.

Yes, he comes from the STM32 world but you can't hardly get your hands on STM32 these days so I switched to the ATtiny for this project.
